doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1446 @@
+NatOrch listens on the notifications of the failed NAT entries from the Syncd.
+For the failed NAT entries, NatOrch removes the corresponding entries from the internal cache and from the conntrack table. That will remove the entry from the APP_DB.
+
+### 3.3.3.4 Clear command
+When the administrator issues the clear command, NatOrch flushes the conntrack table, which in turn results in deleting the entries in the APP_DB by the Natsyncd daemon.
+
+### 3.3.3.5 Max limit on the NAT entries
+NatOrch keeps track of the number of static + dynamic entries created and limits the NAT'ted conntrack entries in the kernel to the supported max limit in the hardware. This is done by removing any new conntrack connections that are being created beyond the maximum limit.
+ 
+### 3.3.3.5 Block diagram
+The high level block diagram for NAT in SONiC is captured below.
+ 
+![NAT block diagram](images/Nat_block_diagram.png)
+ 
+## 3.4 Linux Integration
+### 3.4.1 IPtables
+In SONiC, the core NAT logic which involves the (IP + port) translation dynamically is the iptables module in the kernel.
+The iptables uses the Netfilter hooks in the kernel to execute its NAT'ting rules at different points in the packet forwarding path.
+
+The important hooks that iptables registers with Netfilter are the ones for PREROUTING and POSTROUTING. The DNAT rules if available are applied in the PREROUTING chain and SNAT rules if available are applied in the POSTROUTING chain.
+
+The packets that are going from one zone to another zone in the hardware are trapped to CPU as NAT miss packets if there are no matching NAT entries in the hardware. Such packets are received by the KNET kernel module and passed on to the netdev interface (corresponding to the interface the packet came on) in the Network stack.
+
+Before the L3 forwarding is done in the Kernel, the iptables rules in PREROUTING chain are applied. If there are any DNAT rules that the packets match against, they translate the DIP and L4 port to the IP and L4 port as applicable. In this process, the DNAT state and DIP of the connection entry in the connection tracking table are updated.
+
+After the L3 forwarding is done and before the packet is about to be sent out, the rules in the POSTROUTING chain are applied which do the SNAT (translate the SIP + L4 port to the public IP + L4 port). In this process, the SNAT state and SIP of the connection entry in the connection tracking table are updated.
+
+As long as the hardware entry is not installed, the NAT translation and forwarding is done in the Linux kernel.
+
+### 3.4.2 Connection tracking
+Connection tracking module in the kernel creates the connection entries and maintains their states as and when the packet traverses the forwarding path in the kernel. It keeps track of all the connections created in the system. IPtables module consults the connections tuples tracked in the system during NAT process and updates the connections.
```

Comment:
I feel we need to first describe the asic limitation, 3-tuple matching instead of 5-tuple matching for nat. The limitation creates mismatch between the asic nat model and linux traditional conntrack model. Then, our design should address such mismatch.

Reply:
Done.